src/ultk/effcomm/information_bottleneck/ba.py

```diff
@@ -0,0 +1,228 @@
+        disable_tqdm: bool = False,
+        **kwargs,
+    ) -> list:
+
+        results = []
+        betas = np.sort(self.betas)
+
+        if reverse:
+            # sort betas in decreasing order
+            betas = betas[::-1]
+            # start with bijective mapping
+            init_q = np.eye(len(self.ln_px))
+        else:
+            # Random degenerate initialization
+            xhat = random_stochastic_matrix(shape=(1, self.xhat_size), gamma=1e-2)
+            init_q = np.stack([xhat.squeeze()] * len(self.ln_px))
+
+        pbar = tqdm(
+            betas,
+            disable=disable_tqdm,
+            desc=f"{'reverse ' if reverse else ''}annealing beta",
+        )
+        for beta in pbar:
+            pbar.set_description(f"beta={beta:.5f}")
+            candidates = []
+            for _ in range(num_restarts + 1):
+                self.blahut_arimoto(beta, *args, init_q=init_q, **kwargs)
+                cand = self.results[-1]
+                init_q = cand.qxhat_x
+                candidates.append(cand)
+            best = min(candidates, key=lambda x: x.rate + beta * x.distortion)
+            results.append(best)
+
+        if reverse:
+            results = results[::-1]
+
+        return results
+
+    ############################################################################
+    # Blahut Arimoto iteration
+    ############################################################################
+
+    def blahut_arimoto(
+        self,
+        beta,
+        *args,
+        **kwargs,
+    ) -> None:
+        """Update the self-consistent equations for a Rate Distortion objective.
+
+        Args:
+            beta: (scalar) the slope of the rate-distoriton function at the point where evaluation is required
+        """
+        len_x = len(self.ln_px)
+        if "init_q" in kwargs:
+            init_q: np.ndarray = kwargs["init_q"]
+            # Add small value for working in logspace
+            # init_q_precision = init_q + PRECISION
+            init_q_precision = add_noise_to_stochastic_matrix(init_q, weight=1e-2)
+            init_q_precision /= init_q_precision.sum(axis=1, keepdims=True)
+            self.ln_qxhat_x = np.log(init_q_precision)
+        else:
+            self.ln_qxhat_x = np.log(random_stochastic_matrix((len_x, len_x)))
+
+        it = 0
+        converged = False
+        while not converged:
+            it += 1
+            prev_q = np.copy(self.ln_qxhat_x)
+
+            # Main BA update
+            self.update_eqs(beta, *args, **kwargs)
+
+            # convergence check
+            # TODO: consider updating the Result tuple to include convergence field. if converged, record the iteration. If not, record False/None
+            if self.ignore_converge:
+                converged = it >= self.max_it
+            else:
+                converged = (
+                    it == self.max_it
+                    or np.sum(np.abs(np.exp(self.ln_qxhat_x) - np.exp(prev_q)))
```

Comment:
I'd prob just express the bound in log-space instead of exp'ing, but not a huge deal